src/irmin-pack/unix/chunked_suffix.ml

```diff
@@ -0,0 +1,189 @@
+      in
+      try Ok (v chunk_num create_chunk)
+      with OpenInventoryError err ->
+        Error (err : open_error :> [> open_error ])
+  end
+
+  type t = { inventory : Inventory.t }
+
+  (* A lightweight wrapper around creating/opening append only files as chunks *)
+  module Ao_chunk = struct
+    let path = Layout.V4.suffix_chunk
+
+    let create_rw ~root ~chunk_idx ~overwrite ~auto_flush_threshold
+        ~auto_flush_procedure =
+      let path = path ~root ~chunk_idx in
+      Ao.create_rw ~path ~overwrite ~auto_flush_threshold ~auto_flush_procedure
+
+    let open_rw ~root ~chunk_idx ~auto_flush_threshold ~auto_flush_procedure
+        ~end_poff ~dead_header_size =
+      let path = path ~root ~chunk_idx in
+      Ao.open_rw ~path ~end_poff ~dead_header_size ~auto_flush_threshold
+        ~auto_flush_procedure
+
+    let open_ro ~root ~chunk_idx ~end_poff ~dead_header_size ~is_appendable =
+      let open Result_syntax in
+      let path = path ~root ~chunk_idx in
+      (* Passing [end_poff] is required by Ao for its consistency check (and
+         only applies to the appendable chunk).
+         TODO: We could change its API in the future so we can not do
+         [Io.size_of_path], which is admittedly very silly.*)
+      let* end_poff =
+        if is_appendable then Ok end_poff else Io.size_of_path path
+      in
+      Ao.open_ro ~path ~end_poff ~dead_header_size
+  end
+
+  let create_rw ~root ~start_idx ~overwrite ~auto_flush_threshold
+      ~auto_flush_procedure =
+    let open Result_syntax in
+    let chunk_idx = start_idx in
+    let+ ao =
+      Ao_chunk.create_rw ~root ~chunk_idx ~overwrite ~auto_flush_threshold
+        ~auto_flush_procedure
+    in
+    let chunk = { idx = chunk_idx; start_off = Int63.zero; ao } in
+    let inventory = Inventory.v 1 (Fun.const chunk) in
+    { inventory }
+
+  let open_rw ~root ~end_poff ~start_idx ~chunk_num ~dead_header_size
+      ~auto_flush_threshold ~auto_flush_procedure =
+    let open Result_syntax in
+    let open_chunk ~chunk_idx ~is_appendable =
+      match is_appendable with
+      | true ->
+          Ao_chunk.open_rw ~root ~chunk_idx ~end_poff ~auto_flush_threshold
+            ~auto_flush_procedure ~dead_header_size
+      | false ->
+          Ao_chunk.open_ro ~root ~chunk_idx ~end_poff ~dead_header_size
+            ~is_appendable
+    in
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let open_ro ~root ~end_poff ~dead_header_size ~start_idx ~chunk_num =
+    let open Result_syntax in
+    let open_chunk = Ao_chunk.open_ro ~root ~end_poff ~dead_header_size in
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let appendable_ao t = (Inventory.appendable t.inventory).ao
+  let end_poff t = appendable_ao t |> Ao.end_poff
+
+  let read_exn t ~off ~len buf =
+    let chunk = Inventory.find ~off t.inventory in
+    let poff = chunk_off_to_poff chunk off in
```

Comment:
I actually like that a chunked file is an alternative abstraction for implementing an append-only file, it looks very clean not to leak the chunking trick elsewhere.
However, I would prefer if the abstraction was not in charge of choosing the chunk filenames (for example by receiving the list of chunk filenames as an argument to `open_*` from the file manager) But we could clean this in a parallel PR in the interest of having @metanivek move forward with the MVP!

Reply:
> I actually like that a chunked file is an alternative abstraction for implementing an append-only file, it looks very clean not to leak the chunking trick elsewhere.

You've pinpointed the great advantage of this approach: a patch that feels natural and that moves the minimal amount of dirt. This seems like the the best approach right now, but...

When you say "elsewhere" you are in fact referring to the dispatcher/file_manager pair that are abstractions under our control that have been written very recently for that exact purpose. 

The situation on master is: The file_manager holds the file descriptor and the dispatcher chooses a file descriptor and an offset. Cheating with that [KISS](https://en.wikipedia.org/wiki/KISS_principle) fact will increase the cognitive complexity of irmin-pack. If we preserve that situation I am confident that we will have no surprise when implementing the lower layer.

To illustrate that, picture a new comer to the codebase that is on the path to learning how irmin-pack updates and syncs with the file-system:
- They will find a dispatcher abstraction that doesn't take care of all the dispatching.
- On top of the global offsets and the physical offset they will find a third type of offset which corresponds to the conversion from global to legacy-suffix, just before the conversion from legacy-suffix to physical.

Reply:
@Ngoguey42 thanks for the feedback and historical perspective. I did consider putting this code across the File Manager and Dispatcher but (as is evidenced by this PR :smile:) decided against that path. Before responding to some of the above points, here is more context on my thought process.

I think it is important for us to capture the conceptual pieces of software as self-contained modules with good interfaces. Spreading the concept/logic/usage for a chunked suffix across two (or likely more when you involve `Pack_store`'s usage of `Suffix`) modules does not feel optimal to me. This is probably the most important and motivating reason for me creating a `Chunked_suffix` module. Why is this important to me? I believe that identifying the important concepts and modelling them with good interfaces is one of the most important tasks when trying to build maintainable software that can change over time. I think chunked suffix is important enough that it has its own interface and module.

As I look at the two files mentioned, File Manager and Dispatcher, here is how I see the distinction and how I think a little bit about how this code might evolve in the future:

- The File Manager is the container/orchestrating entity for the conceptual files (Control, Dict, Prefix, Suffix, Index, etc.) that compose an `irmin-pack` store. It is responsible for opening, closing, flushing, swapping, etc.. It also does some reading and writing of these files but: 
	- reading from "the pack file" is coordinated by the Dispatcher 
	- writing also happens directly to the Suffix (eg in `Pack_store`).
- The Dispatcher is the router for reading from the pack file (verbosely Dispatcher is `Pack_file_router` in my mind). Today, the pack file conceptually is a Prefix and a Suffix (independently of how the suffix is implemented). This PR does not change the routing logic between these two concepts. This is one of the strengths (from my perspective) of the approach -- the PR merely replaces how the concept of Suffix is implemented with minimal changes to how you create or interact with it. In the future, I could see us adding a `Vault`/`Lower`/`Whatever` concept for the archive store that the Dispatcher is also responsible for routing to when reading data.

---

I think the above addresses some of the comments but I'd like to reiterate/respond directly, too.

> If we preserve that situation I am confident that we will have no surprise when implementing the lower layer.
For a similar motivating reason as `Chunked_suffix`, I think the lower layer is an important enough concept that it will have its own interface/module and then plug into the Dispatcher's routing logic. I don't foresee any difficulties/surprises with this but I'd be interested if you do.

> a new comer to the codebase that is on the path to learning how irmin-pack updates and syncs with the file-system
This is a difficult code base to comprehend and in part I think (I haven't done a detailed analysis this is more of a hunch :salt:) because concepts/details tend to leak across module boundaries. I think File Manager and Dispatcher are a good step towards adding important concepts to the code base. Similar to the work to break apart the Gc module into good stand-alone concepts (and not a junk drawer utils :wink:), I think adding important concepts (like chunked suffix) as stand-alone modules will benefit the code over time and help paths of learning.

> They will find a dispatcher abstraction that doesn't take care of all the dispatching.
The Dispatcher routes between prefix and suffix. It is true the the chunked suffix has its own internal "dispatching" but I don't find this to be a problem since it is in the scope of the chunked suffix itself which makes it easy to reason about and explain (from my perspective).

> On top of the global offsets and the physical offset they will find a third type of offset which corresponds to the conversion from global to legacy-suffix, just before the conversion from legacy-suffix to physical.
Chunks require a new offset mapping, regardless of where it is implemented. The way this PR maps offsets (and how I would think about it even if the code were located say in the Dispatcher) is global offset -> suffix offset -> chunk physical offset. What you are calling `legacy-suffix` is just the conceptual suffix (and perhaps upper suffix in the future). You could maybe avoid the suffix offset if this code lived in the Dispatcher, but I think there is strength and clarity in its presence:
- it fits and reinforces the model of GC splitting the pack file into a prefix and suffix
- when considering `Chunked_suffix` code you only need to think about suffix offset to chunk off/poff mapping 

---

Hopefully the above makes sense and at least gives reason for why I chose the path I did and how I'm considering future changes.

**tl;dr** I think adding important concepts like Chunked Suffix as stand-alone modules with well-defined interfaces is a good architectural decision and will benefit the code as we continue to evolve and improve it. I also think it fits into the current architecture of File Manager and Dispatcher at the conceptual level.

Reply:
@art-w 

> However, I would prefer if the abstraction was not in charge of choosing the chunk filenames

I like this idea. I wasn't quite sure about passing the root. To add to your idea of passing a list of path names: instead of `root : string` the open/create functions could have a `chunk_path : (chunk_idx:int -> string)` parameter. This leaves control with chunked suffix for creating the paths it needs but removes from its control where those paths are. That type would also play nicely with `Layout.V4.suffix_chunk ~root`.

Reply:
When I first read the codebase, I missed a similar abstraction for "sparse file" to clarify that the mapping file is only used with a compacted object file to implement a "read-only file" interface (and that the mapping doesn't overlap with the suffix). Currently this is conflated in the dispatcher and I think it doesn't scale to also have the chunks logic in there... But the dispatcher could only be in charge of dispatching left to the sparse prefix file, or right to the chunked suffix based on the split offset.

Reply:
Hm well, a `string list` or a `chunk_num:int * chunk_path:(int -> string)` are isomorphic for your purpose :P but the list seems easier to read in the signature and with no "out of bounds" risk! (This is only shifting a `List.init` to the file manager)

Reply:
It is a bit like comparing direct array/list construction with `(Array|List).init`. As a hunch I prefer the latter pattern here to leave control with the chunked suffix, but I need to think more about why I have this preference. I think I could likely be happy with either change. I also have a hunch that this particular point might be clear how to resolve once we have split and gc in place since we will have a complete picture of how the chunked suffix works. Good news: we have two isomorphic iterations to entertain! :smile:

Reply:
(I'm adding my two cents here, but with no intention of blocking this PR, feel free to ignore :) )

I agree with @Ngoguey42 on this, I would much prefer to have the dispatching logic between files be entirely in the dispatcher. There are several bits of logic that I find hard to track, and I think we should try to have them in a single place: 
- physical/virtual offsets conversions;
- reads/writes that lead to errors (due to file boundaries)
- converting read/write to IO.read/write (more abstractions we have, more indirection too).

> Spreading the concept/logic/usage for a chunked suffix across two (or likely more when you involve Pack_store's usage of Suffix) 

It should be that all code above the dispatcher use the dispatcher's read and write, the fact that the suffix leaks in the pack store, I think is something to fix.

Reply:
We discussed this offline with @metanivek. I would TL;DR our discussion this way: I believe that what lands on the main branch should be a somewhat definitive version of the code that everyone agrees on while your approach is more iterative.

Let's do it your way.